File: server/controllers/order_controller.py
from fastapi import HTTPException
from bson import ObjectId
from datetime import datetime
import time
import random
from typing import List

from database import get_database
from models.order import Order, OrderCreate, PaymentStatus

async def create_order(order_data: OrderCreate) -> Order:
    """Create a new order with user details"""
    db = get_database()
    
    # Validate book exists and has stock
    book_id = order_data.book_id
    if not ObjectId.is_valid(book_id):
        raise HTTPException(status_code=400, detail="Invalid book ID")
    
    book = await db.books.find_one({"_id": ObjectId(book_id)})
    if not book:
        raise HTTPException(status_code=404, detail="Book not found")
    
    if book.get("stock", 0) < 1:
        raise HTTPException(status_code=400, detail="Book out of stock")
    
    # Generate unique order ID
    order_id = f"ORD{int(time.time())}{random.randint(100, 999)}"
    
    # Calculate total amount
    delivery_charges = 50
    total_amount = book["price"] + delivery_charges
    
    # Create order document
    order_dict = {
        "bookId": ObjectId(book_id),
        "userDetails": order_data.user_details.model_dump(by_alias=True),
        "orderId": order_id,
        "amount": book["price"],
        "deliveryCharges": delivery_charges,
        "totalAmount": total_amount,
        "paymentStatus": PaymentStatus.PENDING.value,
        "paymentId": None,
        "razorpayOrderId": None,
        "deliveryDate": None,
        "paymentSignature": None,
        "createdAt": datetime.utcnow(),
        "updatedAt": datetime.utcnow()
    }
    
    result = await db.orders.insert_one(order_dict)
    created_order = await db.orders.find_one({"_id": result.inserted_id})
    
    return created_order

async def get_order_by_id(order_id: str) -> Order:
    """Get order by MongoDB ID"""
    if not ObjectId.is_valid(order_id):
        raise HTTPException(status_code=400, detail="Invalid order ID")
    
    db = get_database()
    
    # Aggregate to populate book details
    pipeline = [
        {"$match": {"_id": ObjectId(order_id)}},
        {
            "$lookup": {
                "from": "books",
                "localField": "bookId",
                "foreignField": "_id",
                "as": "bookId"
            }
        },
        {"$unwind": "$bookId"}
    ]
    
    orders = await db.orders.aggregate(pipeline).to_list(1)
    
    if not orders:
        raise HTTPException(status_code=404, detail="Order not found")
    
    return orders[0]

async def get_all_orders() -> List[Order]:
    """Get all orders with populated book details"""
    db = get_database()
    
    pipeline = [
        {
            "$lookup": {
                "from": "books",
                "localField": "bookId",
                "foreignField": "_id",
                "as": "bookId"
            }
        },
        {"$unwind": "$bookId"},
        {"$sort": {"createdAt": -1}}
    ]
    
    orders = await db.orders.aggregate(pipeline).to_list(100)
    return orders

async def get_order_by_order_id(order_id: str) -> Order:
    """Get order by orderId field"""
    db = get_database()
    
    pipeline = [
        {"$match": {"orderId": order_id}},
        {
            "$lookup": {
                "from": "books",
                "localField": "bookId",
                "foreignField": "_id",
                "as": "bookId"
            }
        },
        {"$unwind": "$bookId"}
    ]
    
    orders = await db.orders.aggregate(pipeline).to_list(1)
    
    if not orders:
        raise HTTPException(status_code=404, detail="Order not found")
    
    return orders[0]

File: server/controllers/orderController.js
import Order from '../models/Order.js';
import Book from '../models/Book.js';

// Create order with user details
export const createOrder = async (req, res) => {
  try {
    const { bookId, userDetails } = req.body;

    // Validate book exists and has stock
    const book = await Book.findById(bookId);
    if (!book) {
      return res.status(404).json({
        success: false,
        message: 'Book not found'
      });
    }

    if (book.stock < 1) {
      return res.status(400).json({
        success: false,
        message: 'Book out of stock'
      });
    }

    // Generate unique order ID
    const orderId = `ORD${Date.now()}${Math.floor(Math.random() * 1000)}`;

    // Calculate total amount
    const deliveryCharges = 50;
    const totalAmount = book.price + deliveryCharges;

    const order = await Order.create({
      bookId,
      userDetails,
      orderId,
      amount: book.price,
      deliveryCharges,
      totalAmount,
      paymentStatus: 'Pending'
    });

    res.status(201).json({
      success: true,
      message: 'Order created successfully',
      data: order
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Error creating order',
      error: error.message
    });
  }
};

// Get order by ID
export const getOrderById = async (req, res) => {
  try {
    const order = await Order.findById(req.params.id).populate('bookId');

    if (!order) {
      return res.status(404).json({
        success: false,
        message: 'Order not found'
      });
    }

    res.status(200).json({
      success: true,
      data: order
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Error fetching order',
      error: error.message
    });
  }
};

// Get all orders (Admin)
export const getAllOrders = async (req, res) => {
  try {
    const orders = await Order.find().populate('bookId').sort({ createdAt: -1 });

    res.status(200).json({
      success: true,
      data: orders
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Error fetching orders',
      error: error.message
    });
  }
};
